src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs

```diff
@@ -0,0 +1,843 @@
+                var blittableToDynamicWrapper = new ReduceMapResultsOfStaticIndex.DynamicIterationOfAggregationBatchWrapper();
+                blittableToDynamicWrapper.InitializeForEnumeration(_result.Results);
+
+                var results = new List<object>();
+                IPropertyAccessor propertyAccessor = null;
+                foreach (var output in reducingFunc(blittableToDynamicWrapper))
+                {
+                    propertyAccessor ??= PropertyAccessor.Create(output.GetType(), output);
+                    results.Add(output);
+                }
+
+                if (propertyAccessor == null)
+                {
+                    // TODO: test this
+                    _result.Results.Clear();
+                    return;
+                }
+
+                var objects = new AggregatedAnonymousObjects(results, propertyAccessor, _context, djv =>
+                {
+                    // TODO: handle metadata merge
+                    // TODO: score, distance
+
+                    djv[Constants.Documents.Metadata.Key] = DummyDynamicJsonValue;
+                });
+
+                _result.Results = objects.GetOutputsToStore().ToList();
+            }
+
+            private void ReduceDynamicResults()
+            {
+                var grouping = new Dictionary<GroupKey, List<BlittableJsonReaderObject>>();
+                foreach (BlittableJsonReaderObject item in _result.Results)
+                {
+                    var parts = new List<object>();
+                    foreach (var fld in _query.Metadata.GroupBy)
+                    {
+                        if (item.TryGet(fld.Name, out object val) == false)
+                        {
+                            throw new InvalidOperationException("Unable to reduce sharded query, missing field in projection: " + fld.Name);
+                        }
+
+                        parts.Add(val);
+                    }
+
+                    grouping.GetOrAdd(new GroupKey(parts)).Add(item);
+                }
+                
+                _result.Results.Clear();
+                
+                foreach (var (key, items) in grouping)
+                {
+                    //TODO: handle metadata merge
+                    var result = new DynamicJsonValue();
+                    for (int i = 0; i < _query.Metadata.GroupBy.Length; i++)
+                    {
+                        var groupByField = _query.Metadata.GroupBy[i];
+                        result[groupByField.Alias ?? groupByField.Name] = key[i];
+                    }
+
+                    foreach (var fld in _query.Metadata.SelectFields)
+                    {
+                        switch (fld.AggregationOperation)
+                        {
+                            // Count & Sum needs to be summed, since we are aggregating results here
+                            case AggregationOperation.Count:
+                            case AggregationOperation.Sum:
```

Comment:
Here we have a special handling and distinction whether the index is fanout or not:

https://github.com/ravendb/ravendb/blob/c8c50a8507f0021c278a28433c668819c7650f7c/src/Raven.Server/Documents/Indexes/MapReduce/Auto/AutoMapReduceIndex.cs#L259-L283

In general it would be great to have single code for that. Currently I think the code here is a bit duplicated.

Reply:
Removed. Using now `AggregateOn` from `ReduceMapResultsOfAutoIndex`